### PR TITLE
allow walk function alter its recursive path

### PR DESCRIPTION
**Pull request by vinsia**
allow walk function alter it's recursive path when status is False

### CODE REVIEW

**Comment by vinsia** [comment]
@mtth please take a look.

**Comment by vinsia** [comment]
for os.walk, this is a common usage. here is the documents.

> When topdown is True, the **caller can modify the dirnames list in-place** (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames; this can be used to prune the search, impose a specific order of visiting, or even to inform walk() about directories the caller creates or renames before it resumes walk() again. Modifying dirnames when topdown is False has no effect on the behavior of the walk, because in bottom-up mode the directories in dirnames are generated before dirpath itself is generated.

also i'm facing the same problem, so i think it should be so.

**Comment by mtth** [comment]
Interesting, then it's probably worth including. Can you add a test and guard it behind an option (e.g. `allow_dir_changes`) for backwards-compatibility?

File: hdfs/client.py [hunk 1]
```diff
@@ -1049,11 +1049,14 @@ def _walk(dir_path, dir_status, depth):
...
       else:
+        _dir_infos = [name for name, _ in dir_infos]
         yield (
           dir_path,
-          [name for name, _ in dir_infos],
+          _dir_infos,
           [name for name, _ in file_infos],
         )
+        dir_infos_map = dict((name, s) for name, s in dir_infos)
+        dir_infos = [(name, dir_infos_map[name]) for name in _dir_infos]
```
Comment (mtth) [inline hdfs/client.py]: It sounds like `os.walk` also supports adding directories. If that's the case, we should probably do the same here (adding logic to fetch any missing status).

File: hdfs/client.py [hunk 2]
```diff
@@ -1049,11 +1049,14 @@ def _walk(dir_path, dir_status, depth):
       if status:
         yield ((dir_path, dir_status), dir_infos, file_infos)
       else:
+        _dir_infos = [name for name, _ in dir_infos]
```
Comment (mtth) [inline hdfs/client.py]: `dir_names`

**Comment by vinsia** [comment]
@mtth addressed, PTAL

**Comment by mtth** [comment]
Thanks @vinsia; released in `2.5.7`.

**Pull request merged**